### PR TITLE
Added `_shards` header to all write responses.

### PR DESCRIPTION
**Pull request by martijnvg**
The header indicates to how many shard copies (primary and replicas shards) a write was supposed to go to, to how many shard copies to write succeeded and potentially captures shard failures if writing into a replica shard fails.

For async writes it also includes the number of shards a write is still pending.

### CODE REVIEW

**Comment by martijnvg** [comment]
@bleskes @javanna I applied the feedback and Boaz's commits (https://github.com/bleskes/elasticsearch/commit/a9b659d1c504ca2f3ec1416ceb3d01be9e4b81aa and https://github.com/bleskes/elasticsearch/commit/61804179be69c8f5e723e5da28fc07b35474f080), rebased to master and squashed everything to a single commit for clarity purposes. 

I think this is getting close to get merged now, would be great if someone else can take a look at the PR before it gets merged in.

File: docs/reference/docs/delete-by-query.asciidoc [hunk 1]
```diff
@@ -32,9 +32,8 @@ commands is:
     "_indices" : {
         "twitter" : { 
             "_shards" : {
-                "total" : 5,
-                "successful" : 5,
-                "failed" : 0
+                "total" : 10,
+                "successful" : 10,
```
Comment (imotov) [inline docs/reference/docs/delete-by-query.asciidoc]: +1 on adding `failed` back to docs.

File: src/test/java/org/elasticsearch/deleteByQuery/DeleteByQueryTests.java [hunk 2]
```diff
@@ -194,4 +189,14 @@ public void testDeleteByTermQuery() throws ExecutionException, InterruptedExcept
     private static String indexOrAlias() {
         return randomBoolean() ? "test" : "alias";
     }
+
+    private void assertSyncShardInfo(ActionWriteResponse.ShardInfo shardInfo, NumShards numShards) {
+        assertThat(shardInfo.getTotal(), equalTo(numShards.totalNumShards));
+        assertThat(shardInfo.getSuccessful(), greaterThanOrEqualTo(1));
```
Comment (imotov) [inline src/test/java/org/elasticsearch/deleteByQuery/DeleteByQueryTests.java]: Why not  `greaterThanOrEqualTo(numShards.numPrimaries)` or even `equalTo(numShards.totalNumShards)?

File: src/main/java/org/elasticsearch/action/deletebyquery/ShardDeleteByQueryResponse.java [hunk 3]
```diff
@@ -19,24 +19,45 @@
...
+    private ShardId shardId;
+
+    public ShardDeleteByQueryResponse() {
+    }
+
+    public ShardDeleteByQueryResponse(ShardId shardId) {
+        this.shardId = shardId;
+    }
+
+    public ShardId getShardId() {
```
Comment (imotov) [inline src/main/java/org/elasticsearch/action/deletebyquery/ShardDeleteByQueryResponse.java]: This shardId seems redundant to me as well.

File: src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 4]
```diff
@@ -378,18 +377,20 @@ private void setResponse(BulkItemRequest request, BulkItemResponse response) {
...
             this.response = response;
             this.mappingTypeToUpdate = mappingTypeToUpdate;
             this.op = op;
         }
 
         @SuppressWarnings("unchecked")
-        <T> T response() {
+        <T extends ActionWriteResponse> T response() {
+            // set all to 0, we will embed this into the replica request and not use it
+            response.setShardInfo(new ActionWriteResponse.ShardInfo(0, 0, 0));
```
Comment (imotov) [inline src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: That makes me think that ShardInfo needs a public default constructor.

**Comment by imotov** [comment]
I left a couple of small comments. In general looks good to me.

File: src/test/java/org/elasticsearch/deleteByQuery/DeleteByQueryTests.java (hunk 2 above)
Comment (martijnvg) [inline src/test/java/org/elasticsearch/deleteByQuery/DeleteByQueryTests.java]: I don't know why this wasn't using `numShards.numPrimaries`, let me change that!

**Comment by martijnvg** [comment]
@imotov I updated the PR and applied your comments.

**Comment by imotov** [comment]
LGTM

**Pull request merged**